### PR TITLE
Detection of invalid regions.

### PR DESCRIPTION
**Pull request by hrittich**
When setting Dirichlet boundary conditions invalid region ids were silently
ignored. This will now raise an exception.

### CODE REVIEW

**Comment by dham** [comment]
I'm not convinced this is the Right Thing. Surely it's valid to run a model on a domain where some of the region ids are not there (ie are zero sized).

**Comment by dorugeber** [comment]
I think there's a difference between a "zero-sized" set and a "non-existent" set.

If a user is running a program on a range of input meshes where a mesh marker may or may not exist, I think I'd prefer them to explicitly (apply the BCs/do an integral) conditional on the presence/absence of the mesh marker.

Maybe a visible warning is better than an exception, but Firedrake is already 'noisy' and the warning could easily be lost in the midst of other output.

**Comment by dham** [comment]
Hmm. I guess the user can always catch the exception. OK...

**Comment by dorugeber** [comment]
My last para was supposed to say "a warning is better than nothing", rather than "a warning is better than an exception".

-No, I don't want the user to catch the exception and change their form.  I'd just prefer them to write
```
bcs = [...]
if 3 in mesh.markers:
    bc.append(DirichletBC(V, Expression(...), 3))
```
or (I guess)
```
a = ...
if 3 in mesh.markers:
    a += u*v*dx(3)
```
instead of doing things unconditionally and having Firedrake silently discard some operations.

**Comment by wence-** [comment]
I don't think this is right, or at least, I'm worried about what will go on in parallel where some processes don't have markers but others do.  If I read the code correctly I think this will end up with some processes returning a lookup error and others not.  Is the right thing to do to check in the set of unique_markers?

**Comment by dorugeber** [comment]
Oh, yeah, I notice the Travis build stalled on `tests/regression/test_poisson_mixed_strong_bcs.py::test_poisson_mixed_parallel_fieldsplit `

File: firedrake/mesh.py [hunk 1]
```diff
@@ -106,6 +106,15 @@ def subset(self, markers):
         try:
             return self._subsets[markers]
         except KeyError:
+            # check that the given markers are valid
+            for marker in markers:
+                if marker not in self.markers:
```
Comment (wence-) [inline firedrake/mesh.py]: I think the parallel issue goes away if this tests in `self.unique_markers` rather than `self.markers`.

Comment (hrittich) [inline firedrake/mesh.py]: I have tested this. In

    tests/extrusion/test_steady_advection_3D_extr.py

the function *test_near_to_far* fails: "3 is not a valid marker"

I also tested the following. I included

    if markers is not None:
        for marker in markers:
             assert(marker in unique_markers)

 into *__init__* of the *_Facets* class. Which also fails. Thus *unique_markers* seems not always to be a superset of *markers*.

Comment (wence-) [inline firedrake/mesh.py]: OK, the problem here is that the Facets instantiated on extruded meshes didn't supply the unique_markers argument /at all/.

What happens if you include this patch as well?

```diff
@@ -767,7 +775,8 @@ class ExtrudedMesh(Mesh):
                                         "exterior",
                                         exterior_f.facet_cell,
                                         exterior_f.local_facet_number,
-                                        exterior_f.markers)
+                                        exterior_f.markers,
+                                        unique_markers=exterior_f.unique_markers)
 
         self.ufl_cell_element = ufl.FiniteElement("Lagrange",
                                                   domain=mesh.ufl_cell(),
```

**Comment by hrittich** [comment]
With that changes, the tests run successfully. Although, there were two mistakes in the tests that accessed wrong region IDs.

File: firedrake/mesh.py [hunk 2]
```diff
@@ -49,6 +49,11 @@ def __init__(self, mesh, classes, kind, facet_cell, local_facet_number, markers=
 
         self.local_facet_number = local_facet_number
 
+        # assert that markers is a proper subset of unique_markers
+        if markers is not None:
+            for marker in markers:
+                assert(marker in unique_markers)
```
Comment (wence-) [inline firedrake/mesh.py]: Can you add a message to this assertion?  `assert foo, "Message if assertion failed"`

**Comment by wence-** [comment]
Thanks, a test for the LookupError would be good:

```python
def test_invalid_marker_raises_error():
    with pytest.raises(LookupError):
        code_accessing_invalid_marker
```

**Comment by hrittich** [comment]
I can not figure out, why [this test](https://travis-ci.org/firedrakeproject/firedrake/jobs/71043389) fails. I can not reproduce the failure on my machine.

It looks like that a singular matrix is created. On my machine the matrix generated with and without the patch here are identical...

**Comment by dorugeber** [comment]
It's unrelated to your change, and we've fixed it since yesterday.

**Comment by wence-** [comment]
Thanks, I've merged this, barring the final back-merge from master.

**Pull request closed without merging**